cpp/unit/selection.cpp

```diff
@@ -0,0 +1,202 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include <cmath>
+
+#include "../coord/tile.h"
+#include "../coord/tile3.h"
+#include "../terrain/terrain.h"
+#include "selection.h"
+#include "unit.h"
+
+namespace openage {
+
+UnitSelection::UnitSelection()
+	:
+	drag_active{false} {
+}
+
+bool UnitSelection::on_drawhud() {
+	// the drag selection box
+	if (drag_active) {
+		coord::camhud s = start.to_window().to_camhud();
+		coord::camhud e = end.to_window().to_camhud();
+		glLineWidth(1);
+		glColor3f(1.0, 1.0, 1.0);
+		glBegin(GL_LINE_LOOP); {
+			glVertex3f(s.x, s.y, 0);
+			glVertex3f(e.x, s.y, 0);
+			glVertex3f(e.x, e.y, 0);
+			glVertex3f(s.x, e.y, 0);
+		}
+		glEnd();
+	}
+
+	// draw hp bars for each selected unit
+	glLineWidth(3);
+	for (auto u : this->units) {
+		if (u.second.is_valid()) {
+			Unit *unit_ptr = u.second.get();
+			if (unit_ptr->has_attribute(attr_type::hitpoints)) {
+				auto &hp = unit_ptr->get_attribute<attr_type::hitpoints>();
+				float percent = static_cast<float>(hp.current) / static_cast<float>(hp.max);
+				int mid = percent * 28.0f - 14.0f;
+
+				coord::phys3 &pos_phys3 = unit_ptr->location->pos.draw;
+				coord::camhud pos = pos_phys3.to_camgame().to_window().to_camhud();
+				glColor3f(0.0, 1.0, 0.0);
+				glBegin(GL_LINES); {
+					glVertex3f(pos.x - 14, pos.y + 60, 0);
+					glVertex3f(pos.x + mid, pos.y + 60, 0);
+				}
+				glEnd();
+				glColor3f(1.0, 0.0, 0.0);
+				glBegin(GL_LINES); {
+					glVertex3f(pos.x + mid, pos.y + 60, 0);
+					glVertex3f(pos.x + 14, pos.y + 60, 0);
+				}
+				glEnd();
+			}
+		}
+	}
```

Comment:
a good temporary solution, but we really should avoid the stateful gl stuff in the future.. the graphics engine will hopefully support you soon(tm)